# ECON 8210: Quantitative Macroeconomics Homework 1

Prepared by: Mahmut Eymen Akin

28/08/2024

University of Pennsylvania, Philadelphia 

### **Question 1**

#### **Creating a GitHub repo:**

https://github.com/meymenakin/ECON-8210-F24.git

### **Question 2**

#### **Integration:**

In [2]:
#############################
# 0. Housekeeping
#############################
T=100
ρ=0.04
λ=0.02


u(x)=-exp(-x)
f(x)=exp(-ρ*x)*u(1-exp(-λ*x))

#############################
# 1. Quadrature Methods
#############################

#----------------------------
# 1.1. Functions
#----------------------------

# Midpoint Rule Integration Function
function midpoint_rule(f, a, b, n)
    """
    Approximates the integral of f(x) from a to b using the midpoint rule.

    Parameters:
    - f: The function to integrate.
    - a: The start of the interval.
    - b: The end of the interval.
    - n: The number of subintervals.

    Returns:
    - The approximation of the integral.
    """
    h = (b - a) / n  # step size
    total = 0.0

    # Sum up f at the midpoint of each step
    for i in 0:(n-1)
        midpoint = a + h * (i + 0.5)
        total += f(midpoint)
    end

    return h * total
end


# Trapezoid Rule Integration Function
function trapezoid_rule(f, a, b, n)
    """
    Approximates the integral of f(x) from a to b using the trapezoid rule.

    Parameters:
    - f: The function to integrate.
    - a: The start of the interval.
    - b: The end of the interval.
    - n: The number of subintervals.

    Returns:
    - The approximation of the integral.
    """
    h = (b - a) / n  # step size
    total = 0.5 * (f(a) + f(b))  # start with half of the endpoints' values

    # Sum up f at each interior point
    for i in 1:(n - 1)
        x = a + i * h
        total += f(x)
    end

    return h * total
end

# Simpson's Rule Integration Function
function simpsons_rule(f, a, b, n)
    """
    Approximates the integral of f(x) from a to b using Simpson's Rule.

    Parameters:
    - f: The function to integrate.
    - a: The start of the interval.
    - b: The end of the interval.
    - n: The number of subintervals (must be even).

    Returns:
    - The approximation of the integral.
    """
    if n % 2 != 0
        error("n must be even for Simpson's Rule.")
    end

    h = (b - a) / n  # step size
    total = f(a) + f(b)  # start with the endpoints

    # Sum up terms with weights 4 and 2
    for i in 1:(n-1)
        x = a + i * h
        if i % 2 != 0
            total += 4 * f(x)  # odd indices get a weight of 4
        else
            total += 2 * f(x)  # even indices get a weight of 2
        end
    end

    return (h / 3) * total
end

#----------------------------
# 1.2. Calling Results
#----------------------------

# Get the result of the Approximation with Midpoint Rule
@time midpoint_int = midpoint_rule(f, 0, T, 100)
println("Midpoint rule approximation: $midpoint_int")

# Get the result of the Approximation with Trapezoid Rule
@time trapezoid_int = trapezoid_rule(f, 0, T, 100)
println("Trapezoid rule approximation: $trapezoid_int")

# Get the result of the Approximation with Simpson's Rule
@time simpsons_int = simpsons_rule(f, 0, T, 100)
println("Simpson's rule approximation: $simpsons_int")

#############################
# 2. Monte Carlo
#############################

#----------------------------
# 2.1. Function
#----------------------------

using Random

function monte_carlo_integration(f, a, b, n)
    """
    Approximates the integral of f(x) from a to b using Monte Carlo integration.

    Parameters:
    - f: The function to integrate.
    - a: The start of the interval.
    - b: The end of the interval.
    - n: The number of random points.

    Returns:
    - The approximation of the integral.
    """
    # Generate n random points in the interval [a, b]
    total = 0.0
    for _ in 1:n
        x = a + (b - a) * rand()  # random point in [a, b]
        total += f(x)
    end

    # Calculate the average and multiply by (b - a)
    return (b - a) * total / n
end

#----------------------------
# 2.2. Calling Results
#----------------------------
# Get the result of the Approximation with Monte Carlo Integration
@time mc_int = monte_carlo_integration(f, 0, T, 1000)
println("Monte Carlo integration approximation: $mc_int")

  0.022898 seconds (17.18 k allocations: 1.010 MiB, 99.51% compilation time)
Midpoint rule approximation: -18.20703948621344
  0.010327 seconds (6.30 k allocations: 334.219 KiB, 99.29% compilation time)
Trapezoid rule approximation: -18.214497535093813
  0.009763 seconds (6.15 k allocations: 355.312 KiB, 99.26% compilation time)
Simpson's rule approximation: -18.209527039320182
  0.010200 seconds (23.54 k allocations: 918.000 KiB, 97.19% compilation time)
Monte Carlo integration approximation: -17.374835903985538


##### **Method Comparison**

The integral was approximated using several quadrature methods and Monte Carlo integration, with the following results:

	•	Midpoint Rule: -18.2070
	•	Trapezoid Rule: -18.2145
	•	Simpson’s Rule: -18.2095
	•	Monte Carlo Integration: -17.9801

Note that each method offers a slightly different approximation, highlighting the trade-offs between accuracy and computational time. Simpson’s Rule typically provides higher accuracy due to its polynomial approximation, whereas Monte Carlo simulation exhibits a higher error due to its reliance on random sampling. However, it is much more useful for evaluating higher-dimensional integrals. The execution times indicate that trapezoid and midpoint rules are faster, though Simpson’s Rule achieved the most precise approximation among the quadrature methods.


### **Question 3**

#### **Basic Optimization:**

We want to operate:

$$ \min_{x,y} 100(y-x^2)^2+(1-x)^2 $$

In [3]:
#############################
# 0. Housekeeping
#############################
#import Pkg
#Pkg.add("Optim")

using Optim

#Define objective function
function g(v)
    x, y = v  # Unpack the vector v into x and y
    return 100 * (y - x^2)^2 + (1 - x)^2
end

# Initial guess for the variables [x0, y0]
x0 = [0.0, 0.0]

#############################
# 1. Newton-Raphson Method
#############################
# Optimize the function
@time nr_opt = optimize(g, x0, Newton())

# Display the results
println("Newton-Raphson Method Minimum Point: ", Optim.minimizer(nr_opt))
println("Newton-Raphson Method Function Value at Min: ", Optim.minimum(nr_opt))

#############################
# 2. BFGS Method
#############################
# Optimize the function
@time bfgs_opt = optimize(g, x0, BFGS())

# Display the results
println("BFGS Method Minimum Point: ", Optim.minimizer(bfgs_opt))
println("BFGS Method Function Value at Min: ", Optim.minimum(bfgs_opt))

#############################
# 3. Gradient Descent Method
#############################
# Optimize the function
@time gd_opt = optimize(g, x0, GradientDescent())

# Display the results
println("Gradient Descent Method Minimum Point: ", Optim.minimizer(gd_opt))
println("Gradient Descent Method Function Value at Min: ", Optim.minimum(gd_opt))

#############################
# 4. Conjugate Descent Method
#############################
# Optimize the function
@time cd_opt = optimize(g, x0, ConjugateGradient())

# Display the results
println("Conjugate Descent Method Minimum Point: ", Optim.minimizer(cd_opt))
println("Conjugate Descent Method Function Value at Min: ", Optim.minimum(cd_opt))



  1.555784 seconds (3.47 M allocations: 235.464 MiB, 3.65% gc time, 99.97% compilation time)
Newton-Raphson Method Minimum Point: [0.9999999926663182, 0.9999999853326361]
Newton-Raphson Method Function Value at Min: 5.378288834719864e-17
  0.800650 seconds (1.31 M allocations: 89.691 MiB, 17.25% gc time, 99.93% compilation time)
BFGS Method Minimum Point: [0.9999999926033533, 0.9999999852005563]
BFGS Method Function Value at Min: 5.4714165475110124e-17
  0.111184 seconds (135.74 k allocations: 9.071 MiB, 98.85% compilation time)
Gradient Descent Method Minimum Point: [0.9795724831657034, 0.9595023107810468]
Gradient Descent Method Function Value at Min: 0.0004176427123201452
  0.271920 seconds (367.70 k allocations: 25.000 MiB, 99.90% compilation time)
Conjugate Descent Method Minimum Point: [0.9999999926288975, 0.9999999852578547]
Conjugate Descent Method Function Value at Min: 5.43331525129156e-17


##### **Method Comparison**
Note that with the exception of the Gradient Descent method, each optimization method converged more or less to the same point, with Newton-Raphson and BFGS achieving faster convergence lower memory usage due to their use of gradient and curvature information. Steepest descent requires more iterations to converge fully, but it also required the lowest memory usage. Conjugate descent, on the other hand achieves much better results than steepest descent while using significantly less memory than Newton-Raphson and BFGS methods.

### **Question 4**

#### **Computing Pareto Efficient Allocations**

We want to solve the Pareto Efficient allocations of the following model:
- agents, $ i \in \{1, \dots, n\} $
- goods, $ j \in \{1, \dots, m\} $
- preferences: $$ u^i(x) = \sum_{j=1}^m \alpha_j^i \frac{x_j^{1+\omega_j^i}}{1+\omega_j^i} $$
where $ \alpha_j^i > 0>  \omega_j^i $
- endowments, $ e_j^i > 0, \quad \forall i,j$

Then, define $ \textbf{x} = (x^1,\dots, x^n) $ where $x^i = (x_1^i, \dots, x_m^i)$. Then, the Pareto Efficient allocations will be given by the following maximization problem:
$$ \max_{\textbf{x}} \sum_{i=1}^n \lambda^i \sum_{j=1}^m \alpha_j^i \frac{(x_j^i)^{1+\omega_j^i}}{1+\omega_j^i}  $$
subject to$$ \sum_{i=1}^n x_j^i \leq \sum_{i=1}^n e_j^i, \quad \quad \forall j\in \{1, \dots, m\} $$

This yields the following FOC:
$$[x_j^i]: \quad \lambda^i \alpha_j^i (x_j^i)^{\omega_j^i} - \mu_j =0 $$

$$\implies \frac{(x_j^i)^{\omega_j^i}}{(x_j^k)^{\omega_j^k}} = \frac{\lambda^k \alpha_j^k}{\lambda^i \alpha_j^i} \quad \quad \forall i,k\in \{1,\dots,n\} $$
$$\implies x_j^i = \left( \frac{\lambda^k \alpha_j^k}{\lambda^i \alpha_j^i} \right)^{\frac{1}{\omega_j^i}} (x_j^k)^{\frac{\omega_j^k}{\omega_j^i}}  \quad \quad \forall i,k\in \{1,\dots,n\}  $$


In [4]:
#############################
# 0. Housekeeping
#############################

#Pkg.add("JuMP")
#Pkg.add("Ipopt")
#Pkg.add("DataFrames")

using DataFrames
using Random
using JuMP
using Ipopt

# Define endowments matrix (constant across models)
e = [
    2.0 3.0 1.0;   # Agent 1's endowments
    6.0 4.0 12.0;   # Agent 2's endowments
    2.0 3.0 6.0    # Agent 3's endowments
]


n = 3  # Number of agents
m = 3  # Number of goods

#############################
# 1. No Heterogeneity
#############################

#----------------------------
# 1.1. Define Parameters
#----------------------------
α1 = 0.5       # Alpha parameter for each good and agent
ω1 = -0.2      # Omega parameter for each good and agent (mild diminishing returns)
λ1 = 1.0       # Social weight for each agent

#----------------------------
# 1.2. Optimize
#----------------------------

# Initialize the optimization model
nh_model = Model(Ipopt.Optimizer)
set_optimizer_attribute(nh_model, "print_level", 3)


# Define the decision variables (allocations x_j^i)
@variable(nh_model, x1[1:n, 1:m] >= 0)

# Define the objective function (social welfare function)
@objective(nh_model, Max, sum(λ1 * α1 * x1[i, j]^(1 + ω1) / (1 + ω1) for i in 1:n, j in 1:m))

# Add the resource constraints for each good
for j in 1:m
    @constraint(nh_model, sum(x1[i, j] for i in 1:n) == sum(e[i, j] for i in 1:n))
end

# Solve the model
optimize!(nh_model)

println("Optimal allocations with no heterogeneity:")
for i in 1:n, j in 1:m
    println("x[$i, $j] = ", value(x1[i, j]))
end

#######################################
# 2. Heterogeneity across j and i
#######################################

#----------------------------
# 2.1. Define Parameters
#----------------------------
α2 = [
    0.8 1.2 0.5;   # Agent 1's α values for each good
    1.0 1.1 0.7;   # Agent 2's α values for each good
    0.6 1.3 0.9    # Agent 3's α values for each good
]

ω2 = [
    -0.3 -0.5 -0.2;  # Agent 1's ω values for each good
    -0.4 -0.6 -0.3;  # Agent 2's ω values for each good
    -0.2 -0.4 -0.1   # Agent 3's ω values for each good
]

λ2 = [1.0, 0.8, 1.2]  # Social weights for each agent

#----------------------------
# 2.2. Optimize
#----------------------------

# Initialize the optimization model
h_model = Model(Ipopt.Optimizer)
set_optimizer_attribute(h_model, "print_level", 3)

# Define the decision variables (allocations x_j^i)
@variable(h_model, x2[1:n, 1:m] >= 0)

# Define the objective function (social welfare function with full heterogeneity)
@objective(h_model, Max, sum(λ2[i] * α2[i, j] * x2[i, j]^(1 + ω2[i, j]) / (1 + ω2[i, j]) for i in 1:n, j in 1:m))

# Add the resource constraints for each good
for j in 1:m
    @constraint(h_model, sum(x2[i, j] for i in 1:n) == sum(e[i, j] for i in 1:n))
end

# Solve the model
optimize!(h_model)

# Check the solution status
println("Optimal allocations with heterogeneity:")
for i in 1:n, j in 1:m
  println("x[$i, $j] = ", value(x2[i, j]))
end

#######################################
# 3. 10 agents
#######################################
#----------------------------
# 3.1. Define Parameters
#----------------------------
n2 = 10  # Number of agents
m2 = 10  # Number of goods

Random.seed!(123)

α3 = rand(0.5:0.1:1.5, n2, m2)       # Random values for α between 0.5 and 1.5 for each agent-good pair
ω3 = rand(-0.6:0.1:-0.1, n2, m2)      # Random values for ω between -0.6 and -0.1 for each agent-good pair
λ3 = rand(0.8:0.1:1.2, n2)           # Random values for λ between 0.8 and 1.2 for each agent

e3 = rand(5.0:1.0:10.0, n2, m2)  # Random values between 5 and 10 for endowments of each agent for each good

#----------------------------
# 3.2. Optimize
#----------------------------

# Initialize the optimization model
large_model = Model(Ipopt.Optimizer)
set_optimizer_attribute(large_model, "print_level", 3)


# Define the decision variables (allocations x_j^i)
@variable(large_model, x3[1:n2, 1:m2] >= 0)

# Define the objective function (social welfare function with full heterogeneity)
@objective(large_model, Max, sum(λ3[i] * α3[i, j] * x3[i, j]^(1 + ω3[i, j]) / (1 + ω3[i, j]) for i in 1:n2, j in 1:m2))

# Add the resource constraints for each good
for j in 1:m2
    @constraint(large_model, sum(x3[i, j] for i in 1:n2) == sum(e3[i, j] for i in 1:n2))
end

# Solve the model
optimize!(large_model)

# Collect values in a DataFrame format
    allocation_values = [value(x3[i, j]) for i in 1:n2, j in 1:m2]  # Matrix of optimal values
    allocation_df = DataFrame(allocation_values, 
                              :auto)  # Automatically assign column names
    
    # Rename columns to "Good_1", "Good_2", ..., "Good_10"
    rename!(allocation_df, Symbol.("Good_" .* string.(1:m2)))
    # Add row names for agents
    allocation_df.agent = Symbol.("Agent_" .* string.(1:n2))
    select!(allocation_df, [:agent, Symbol.("Good_" .* string.(1:m2))...])

    println("Optimal allocations as a 10x10 table:")
    # Display the DataFrame with scaled font size using PrettyTables
    println(allocation_df)




******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

Total number of variables............................:        9
                     variables with only lower bounds:        9
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        3
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0


Number of Iterations....: 7

      

#### **Discussion**

Note that as heterogeneity is introduced, the allocations change significantly. Also, when the number of agents increased to 10, the number of iterations for the solver doubled. Hence, it became computationally much more challenging.

### **Question 5**

#### **Computing Equilibrium Allocations**

Recall that we have the following FOC in the SPP:
$$[x_j^i]: \quad \lambda^i \alpha_j^i (x_j^i)^{\omega_j^i} - \mu_j =0 $$

$$\implies x_j^i = \left( \frac{\mu_j}{\lambda^i \alpha_j^i} \right)^{\frac{1}{\omega_j^i}} $$

Now, note that the individual i solves the following problem:
$$\max_{x^i} \; u^i(x) = \sum_{j=1}^m \alpha_j^i \frac{(x_j^i)^{1 + \omega_j^i}}{1 + \omega_j^i} $$
subject to
$$ \sum_{j=1}^m p^j x_j^i \leq \sum_{j=1}^m p^j e_j^i
 $$

This yields the following FOC:
$$[x_j^i]: \quad \alpha_j^i (x_j^i)^{\omega_j^i} - \psi_i p^j=0
 $$ 
$$\implies \alpha_j^i (x_j^i)^{\omega_j^i} = \psi_i p^j $$
$$\implies x_j^i = \left(\frac{\psi_i p^j}{\alpha_j^i}\right)^{\frac{1}{\omega_j^i}} $$
Then, by plugging this expression into the BC, we obtain $\forall i\in\{1,\dots,n\} $ : 
$$\sum_{j=1}^m p^j \left(\frac{\psi_i p^j}{\alpha_j^i}\right)^{\frac{1}{\omega_j^i}} = \sum_{j=1}^m p^j e_j^i $$
Also, from the resource constraints, we get that $\forall j \in \{1,\dots, m\} $:
$$\sum_{i=1}^n \left(\frac{\psi_i p^j}{\alpha_j^i}\right)^{\frac{1}{\omega_j^i}} = \sum_{i=1}^n e_j^i $$
Hence, we have m+n unknowns and m+n equations.

In [6]:
############################################
# 1. Heterogeneity across j and i (m=n=3)
############################################

# Initialize the optimization model
p_h_model = Model(Ipopt.Optimizer)
set_optimizer_attribute(p_h_model, "print_level", 3)

# Define the decision variables for prices (p) and ψ values for each agent
@variable(p_h_model, p1[1:m] >= 0.1)     # Prices for each good (lower bound to prevent zero prices)
@variable(p_h_model, ψ1[1:n] >= 0)       # ψ values for each agent (assuming ψ is non-negative)

# Define the objective function as the sum of squared errors
@objective(p_h_model, Min,
    # Budget constraint terms (sum of squared differences for each agent i)
    sum((sum(p1[j] * (ψ1[i] * p1[j] / α2[i, j])^(1 / ω2[i, j]) for j in 1:m) - sum(p1[j] * e[i, j] for j in 1:m))^2 for i in 1:n) +
    # Resource constraint terms (sum of squared differences for each good j)
    sum((sum((ψ1[i] * p1[j] / α2[i, j])^(1 / ω2[i, j]) for i in 1:n) - sum(e[i, j] for i in 1:n))^2 for j in 1:m)
)

# Solve the model
optimize!(p_h_model)

# Check if the solution converged and display the results
    println("Equilibrium prices (p) when m=n=3:")
    for j in 1:m
        println("p[$j] = ", value(p1[j]))
    end
    


Total number of variables............................:        6
                     variables with only lower bounds:        6
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0


Number of Iterations....: 166

                                   (scaled)                 (unscaled)
Objective...............:   2.6635116155224936e-07    2.6635116155224935e+01
Dual infeasibility......:   5.8869897514063397e-11    5.8869897514063395e-03
Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementar

### Question 6

#### **Value Function Iteration**

##### **6.1 Social Planner Problem**

Let $\Pi_\tau$ and $\Pi_z$ denote the Markov transition matrices for their corresponding variables. Then the Social Planner solves:

$$V(k,i_{-1}, z, \tau) = \max_{c, l,i, k^\prime} \left( \log c + 0.2 \log g - \frac{l^2}{2} + \beta \cdot \sum_{z^\prime} \sum_{\tau^\prime} \Pi_z(z,z^\prime) \Pi_\tau(\tau,\tau^\prime) V(k^\prime, i, z^\prime, \tau^\prime)  \right) $$
subject to
$$    c + i = (1 - \tau) w l + r k $$
$$    c + i + g = e^{z} k^{\alpha} l_t^{1-\alpha}$$
$$    k^\prime = (1-\delta) k + \left[ 1 - 0.05 \left( \frac{i}{i_{-1}} - 1 \right)^2 \right] i $$
$$    g = \tau w l
$$

w/ $\beta = 0.97 $, $\alpha = 0.33$, $\delta=0.1$, $\Pi_\tau=\begin{bmatrix}
0.9 & 0.1 & 0 \\
0.05 & 0.9 & 0.05 \\
0 & 0.1 & 0.9 \\
\end{bmatrix} $, and $\Pi_z=\begin{bmatrix}
0.9727 & 0.0273 & 0 & 0 & 0 \\
0.0041 & 0.9806 & 0.0153 & 0 & 0 \\
0 & 0.0082 & 0.9836 & 0.0082 & 0 \\
0 & 0 & 0.0153 & 0.9806 & 0.0041 \\
0 & 0 & 0 & 0.0273 & 0.9727 \\
\end{bmatrix}$.

##### **6.2 Steady State**

In a Recursive Competitive Equilibrium, the household maximizes:
$$ V(k,i_{-1}, z, \tau) = \max_{c, l,i, k^\prime} \left( \log c + 0.2 \log g - \frac{l^2}{2} + \beta \cdot \sum_{z^\prime} \sum_{\tau^\prime} \Pi_z(z,z^\prime) \Pi_\tau(\tau,\tau^\prime) V(k^\prime, i, z^\prime, \tau^\prime)  \right) $$
subject to
$$ c + i = (1 - \tau) w l + r k$$
$$      k^\prime = (1-\delta) k + \left[ 1 - 0.05 \left( \frac{i}{i_{-1}} - 1 \right)^2 \right] i$$

And the prices will be equal to the marginal productivities:
$$ w(k,z) = (1-\alpha)e^{z}\left(\frac{k}{l}\right)^\alpha \tag{1}$$
$$ r(k,z) = \alpha e^{z}\left(\frac{l}{k}\right)^{1-\alpha} \tag{2}$$

The household's problem yields the following FOCs:
$$ [c]:\quad \frac{1}{c}-\lambda =0 \implies \frac{1}{c}=\lambda \tag{3}$$
$$ [l]:\quad -l+\lambda(1-\tau)w =0 \implies l=\lambda(1-\tau)w  $$
$$ \implies \frac{1}{c} = \frac{l}{(1-\tau)w} \tag{4} $$
which is the household's intratemporal equation for labor-consumption decisions.
Furthermore
$$[k^\prime]: \quad \beta \cdot \sum_{z^\prime} \sum_{\tau^\prime} \Pi_z(z,z^\prime) \Pi_\tau(\tau,\tau^\prime) V_k(k^\prime, i, z^\prime, \tau^\prime) - \mu = 0 $$
$$  \beta \cdot \sum_{z^\prime} \sum_{\tau^\prime} \Pi_z(z,z^\prime) \Pi_\tau(\tau,\tau^\prime) V_k(k^\prime, i, z^\prime, \tau^\prime) = \mu \tag{5}$$

And by the Envelope Theorem:
$$V_k(k, i_{-1}, z, \tau) = r\lambda + (1-\delta) \mu $$
$$\implies V_k(k^\prime, i, z^\prime, \tau^\prime)  =  r^\prime\lambda^\prime + (1-\delta) \mu^\prime \tag{6}$$
Then:
$$[i]: \quad \beta \cdot \sum_{z^\prime} \sum_{\tau^\prime} \Pi_z(z,z^\prime) \Pi_\tau(\tau,\tau^\prime) V_i(k^\prime, i, z^\prime, \tau^\prime) - \lambda + \mu \left[ \left(1- 0.05 \left(\frac{i}{i_{-1}}-1\right)^2 \right) - 0.1\frac{i}{i_{-1}} \left(\frac{i}{i_{-1}}-1 \right) \right]=0 $$
which implies:
$$ \beta \cdot \sum_{z^\prime} \sum_{\tau^\prime} \Pi_z(z,z^\prime) \Pi_\tau(\tau,\tau^\prime) V_i(k^\prime, i, z^\prime, \tau^\prime) = \lambda - \mu \left[ \left(1- 0.05 \left(\frac{i}{i_{-1}}-1\right)^2 \right) + 0.1\frac{i}{i_{-1}} \left(\frac{i}{i_{-1}}-1 \right) \right] \tag{7} $$
Again, by the Envelope Theorem:

$$V_i(k, i_{-1}, z, \tau)= - 0.1 \mu i \left(\frac{i}{i_{-1}}-1 \right)\frac{i}{i^2_{-1}} $$
$$ \implies V_i(k^\prime, i, z^\prime, \tau^\prime) = - 0.1 \mu^\prime i^\prime \left(\frac{i^\prime}{i}-1 \right)\frac{i^\prime}{i^2} \tag{8}$$
Finally, we have the government budget, capital law of motion , and household's budget equation:
$$ g = \tau w l \tag{9}$$ 
$$ k^\prime = (1-\delta) k + \left[1 - 0.05 \left(\frac{i}{i_{-1}}-1\right)^2 \right] i \tag{10}$$
$$ c + i  = (1-\tau)wl + rk \tag{11}$$ 

Hence, we have obtained **11 equations** expressing **11 unknowns** for the *steady state*. Then, the steady-state versions of these equations will become:
$$ w_{ss} = (1 - \alpha) e^{z_{ss}} \left( \frac{k_{ss}}{l_{ss}} \right)^\alpha \tag{1$_{\text{ss}}$} $$
$$ r_{ss} = \alpha e^{z_{ss}} \left( \frac{l_{ss}}{k_{ss}} \right)^{1 - \alpha} \tag{2$_{\text{ss}}$} $$
$$ \frac{1}{c_{ss}} = \lambda_{ss} \tag{3$_{\text{ss}}$}$$
$$ \frac{1}{c_{ss}} = \frac{l_{ss}}{(1 - \tau_{ss}) w_{ss}} \tag{4$_{\text{ss}}$}$$
$$ \beta \cdot V_{k,ss} = \mu_{ss} \tag{5$_{\text{ss}}$} $$
$$ V_{k,ss} = r_{ss}\lambda_{ss} + (1 - \delta) \mu_{ss} \tag{6$_{\text{ss}}$}$$
$$  \beta \cdot V_{i,ss} = \lambda_{ss} -\mu_{ss} \tag{7$_{\text{ss}}$}$$
$$ V_{i,ss} = 0 \tag{8$_{\text{ss}}$}
$$
$$ g_{ss} = \tau_{ss} w_{ss} l_{ss} \tag{9$_{\text{ss}}$}
$$
$$ k_{ss} = (1-\delta) k_{ss} + i \tag{10$_{\text{ss}}$}$$
$$ c_{ss} + i_{ss}  = (1-\tau_{ss})w_{ss}l_{ss} + r_{ss}k_{ss} \tag{11$_{\text{ss}}$}$$ 
Then, further manipulating the equations **($7_{ss}$)** and **($8_{ss}$)** we get $\lambda_{ss} = \mu_{ss}$ and $ V_{i,ss} = 0$. Then, combine this with equation **($3_{ss}$)** to get 8 equations for 8 unknowns in the steady state. Finally, combine **($5_{ss}$)** and **($6_{ss}$)** to reduce the system to 7 equations and 7 unknowns.
$$ w_{ss} = (1 - \alpha) e^{z_{ss}} \left( \frac{k_{ss}}{l_{ss}} \right)^\alpha \tag{1$^{\star}$} $$
$$ r_{ss} = \alpha e^{z_{ss}} \left( \frac{l_{ss}}{k_{ss}} \right)^{1 - \alpha} \tag{2$^{\star}$} $$
$$ \frac{1}{c_{ss}} = \frac{l_{ss}}{(1 - \tau_{ss}) w_{ss}} \tag{3$^{\star}$}$$
$$ \beta (r_{ss}+ 1 - \delta) = 1 \tag{4$^{\star}$} $$
$$ g_{ss} = \tau_{ss} w_{ss} l_{ss} \tag{5$^{\star}$}
$$
$$ k_{ss} = (1-\delta) k_{ss} + i \tag{6$^{\star}$}$$
$$c_{ss} + i_{ss}  = (1-\tau_{ss})w_{ss}l_{ss} + r_{ss}k_{ss} \tag{7$^{\star}$}$$


In [21]:
############################################
# 0. Housekeeping
############################################
using JuMP
using Ipopt

# Define constants
α = 0.33        # Capital share
β = 0.97        # Discount factor
δ = 0.1         # Depreciation rate
τ_ss = 0.25     # Steady-state tax rate
z_ss = 0.0      # Technology level in steady state (log scale)

############################################
# 1. Solve for the Steady State
############################################

# Initialize the JuMP model with the Ipopt solver
ss_model = Model(Ipopt.Optimizer)
set_silent(ss_model) 
# Define the variables with initial guesses
@variable(ss_model, w_ss >= 0 + 0.01)       # Steady-state wage
@variable(ss_model, r_ss >= 0 + 0.01)      # Steady-state return on capital
@variable(ss_model, l_ss >= 0 + 0.01)       # Steady-state labor
@variable(ss_model, c_ss >= 0 + 0.01)       # Steady-state consumption
@variable(ss_model, g_ss >= 0 + 0.01)       # Steady-state government spending
@variable(ss_model, k_ss >= 0 + 0.01)       # Steady-state capital
@variable(ss_model, i_ss >= 0 + 0.01)       # Steady-state investment

# Define the system of equations as constraints
@constraint(ss_model, w_ss == (1 - α) * exp(z_ss) * (k_ss / l_ss)^α)                              # (1★)
@constraint(ss_model, r_ss == α * exp(z_ss) * (l_ss / k_ss)^(1 - α))                              # (2★)
@constraint(ss_model, 1 / c_ss == l_ss / ((1 - τ_ss) * w_ss))                                     # (3★)
@constraint(ss_model, β * (r_ss + 1 - δ) == 1)                                                    # (4★)
@constraint(ss_model, g_ss == τ_ss * w_ss * l_ss)                                                 # (5★)
@constraint(ss_model, k_ss == (1-δ)*k_ss + i_ss)                                                  # (6★)
@constraint(ss_model, c_ss + i_ss == (1 - τ_ss) * w_ss * l_ss + r_ss * k_ss)                      # (7★)

# Solve the system
optimize!(ss_model)

# Extract and print the solution
println("Steady-state solution:")
println("w_ss = ", value(w_ss))
println("r_ss = ", value(r_ss))
println("l_ss = ", value(l_ss))
println("c_ss = ", value(c_ss))
println("g_ss = ", value(g_ss))
println("k_ss = ", value(k_ss))
println("i_ss = ", value(i_ss))

Steady-state solution:
w_ss = 1.0563787126105089
r_ss = 0.13092783636920743
l_ss = 0.930431558829313
c_ss = 0.8515225430103279
g_ss = 0.24572202292143933
k_ss = 3.6975258211514284
i_ss = 0.3697525819524335


##### **6.3 Value Function Iteration with a Fixed Grid**

The **Recursive Competitive Equilibrium** in this economy consists of:
1. A value function $ V(k, i_{-1}, K, I_{-1}, \tau, z) $ for the representative household,
2. Decision rules for consumption $ c(k, i_{-1}, K, I_{-1}, \tau, z) $, labor $ l(k, i_{-1}, K, I_{-1}, \tau, z) $, investment $ i(k, i_{-1}, K, I_{-1}, \tau, z) $, and next-period capital $k'(k, i_{-1}, K, I_{-1}, \tau, z) $,
3. Prices $ w(K, I_{-1}, \tau, z) $ and $ r(K, I_{-1}, \tau, z) $,
4. Government policy $ g(K, I_{-1}, \tau, z) $,
5. Laws of motion for aggregate state variables,

such that:

1. **Household Optimization:**
   Given prices $w(K, I_{-1}, \tau, z) $ and $ r(K, I_{-1}, \tau, z) $, and government policy $ g(K, I_{-1}, \tau, z) $, the household’s value function $ V(k, i_{-1}, K, I_{-1}, \tau, z) $ satisfies:
   $$
   V(k, i_{-1}, K, I_{-1}, \tau, z) = \max_{c, l, i, k'} \left\{ \log c + 0.2 \log g(K, I_{-1}, \tau, z) - \frac{l^2}{2} + 0.97 \mathbb{E} \left[V(k', i, K', I, \tau', z')\right] \right\}
   $$
   subject to:
     $$
     c + i = (1 - \tau) w(K, I_{-1}, \tau, z) l + r(K, I_{-1}, \tau, z) k
     $$
     $$
     k' = 0.9 k + \left[ 1 - 0.05 \left( \frac{i}{i_{-1}} - 1 \right)^2 \right] i
     $$

2. **Firm Optimization:**

   Prices should be equal to marginal productivities of aggregate variables
   - **Wage Rate (Marginal Product of Labor):**
     $$
     w(K, I_{-1}, \tau, z) = 0.67 e^z \left( \frac{K}{L} \right)^{0.33}
     $$
   - **Return on Capital (Marginal Product of Capital):**
     $$
     r(K, I_{-1}, \tau, z) = 0.33 e^z \left( \frac{L}{K} \right)^{0.67}
     $$

3. **Government Budget Constraint:**
   The government sets its consumption $ g(K, I_{-1}, \tau, z) $ to balance the budget each period:
   $$
   g(K, I_{-1}, \tau, z) = \tau w(K, I_{-1}, \tau, z) L
   $$

4. **Market Clearing Conditions and Consistency Conditions:**
   - **Goods Market Clearing:** 
     $$
     \int c(k, i_{-1}, K, I_{-1}, \tau, z) \, d\mu(k, i_{-1}) + I + g(K, I_{-1}, \tau, z) = e^z K^{0.33} L^{0.67}
     $$
   - **Labor Market Clearing:** The aggregate labor supply $ L(S) $ is given by:
     $$
     L(K, I_{-1}, \tau, z) = \int l(k, i_{-1}, K, I_{-1}, \tau, z) \, d\mu(k, i_{-1})
     $$

5. **Law of Motion for Aggregate States:**
   - **Aggregate Capital Consistency:** Aggregate capital in the next period, $ K' $, must match the aggregation of individual capital choices:
     $$
     K'(K, I_{-1}, \tau, z) = \int k'(k, i_{-1}, K, I_{-1}, \tau, z) \, d\mu(k, i_{-1})
     $$
   - **Aggregate Investment Consistency:** Aggregate investment in the previous period, $ I_{-1} $, is consistent with the past period’s investment decisions across individuals:
     $$
     I(K, I_{-1}, \tau, z) = \int i(k, i_{-1}, K, I_{-1}, \tau, z) \, d\mu(k, i_{-1})
     $$


In [33]:
############################################
# 0. Housekeeping
############################################
#import Pkg
#Pkg.add("Interpolations")
#Pkg.add("Parameters")
#Pkg.add("ProgressMeter")
#Pkg.add("Plots")
#Pkg.add("Distributions")
#Pkg.add("Roots")
#Pkg.add("MathOptInterface")

# Import necessary packages
using NLsolve
using LinearAlgebra
using Statistics
using Interpolations
using ProgressMeter
using Plots
using Distributions

# Define model parameters
β = 0.97         # Discount factor
δ = 0.1          # Depreciation rate
α = 0.33         # Capital share
θ = 1 - α        # Labor share
γ = 0.05         # Adjustment cost parameter
ψ = 0.2          # Utility weight on government consumption

# Grids for capital and lagged investment
k_min = value(k_ss) * 0.7
k_max = value(k_ss) * 1.3
k_size = 250
k_grid = collect(range(k_min, k_max, length=k_size))

i_min = value(i_ss) * 0.5
i_max = value(i_ss) * 1.5
i_size = 50
i_grid = collect(range(i_min, i_max, length=i_size))

# Exogenous state variables
tau_states = [0.2, 0.25, 0.3]
Pi_tau = [
    0.9   0.1  0.0;
    0.05  0.9  0.05;
    0.0   0.1  0.9
]

z_states = [-0.0673, -0.0336, 0.0, 0.0336, 0.0673]
Pi_z = [
    0.9727 0.0273 0.0    0.0    0.0;
    0.0041 0.9806 0.0153 0.0    0.0;
    0.0    0.0082 0.9836 0.0082 0.0;
    0.0    0.0    0.0153 0.9806 0.0041;
    0.0    0.0    0.0    0.0273 0.9727
]

# Aggregate variables (assume constant for simplicity)
K = value(k_ss)
I_prev = value(k_ss)

############################################
# 1. Value Function Iteration
############################################

# Initialize value function and policy functions
V = zeros(k_size, i_size, length(tau_states), length(z_states))
c_policy = similar(V)
l_policy = similar(V)
i_policy = similar(V)
w_policy = similar(V)
r_policy = similar(V)
g_policy = similar(V)
kprime_policy = similar(V)

# Define the maximum number of iterations and tolerance for convergence
max_iter = 1000
tol = 1e-6

# Begin Value Function Iteration
println("Starting Value Function Iteration...")

for iter = 1:max_iter
    V_new = copy(V)
    sup_norm = 0.0

    @showprogress for τ_idx = 1:length(tau_states)
        τ = tau_states[τ_idx]
        for z_idx = 1:length(z_states)
            z = z_states[z_idx]

            # Loop over all combinations of k and i_prev
            for (k_i, k) in enumerate(k_grid)
                for (i_prev_i, i_prev) in enumerate(i_grid)
                    # Initialize maximum value
                    V_max = -Inf
                    c_opt = l_opt = i_opt = kprime_opt = w_opt = r_opt = g_opt = NaN

                    # Possible choices for investment i
                    i_choices = range(i_min, i_max, length=20)

                    # Loop over possible choices of i
                    for i in i_choices
                        # Capital accumulation
                        adj_cost = 1 - γ * ((i / i_prev) - 1)^2
                        k_prime = (1 - δ) * k + adj_cost * i

                        # Ensure k_prime is within bounds
                        if k_prime < k_min || k_prime > k_max
                            continue
                        end

                        # Define the system of equations to solve
                        function equations!(F, vars)
                            c = vars[1]
                            l = vars[2]
                            w = vars[3]
                            r = vars[4]

                            # Equation 1: Budget Constraint
                            F[1] = c + i - ((1 - τ) * w * l + r * k)
                            
                            # Equation 2: Labor Supply FOC
                            F[2] = -l + ((1 - τ) * w) / c
                            
                            # Equation 3: Wage Determination
                            F[3] = w - (1 - α) * exp(z) * (K / l)^α
                            
                            # Equation 4: Rental Rate Determination
                            F[4] = r - α * exp(z) * (K / l)^(α - 1)
                        end

                        # Initial guesses for c, l, w, r
                        c_guess = 1.0
                        l_guess = 1.0
                        w_guess = 1.0
                        r_guess = 0.1
                        initial_guesses = [c_guess, l_guess, w_guess, r_guess]

                        # Solve the system
                        sol = nlsolve(equations!, initial_guesses; method=:newton, xtol=1e-8, ftol=1e-8)

                        # Check if the solution is valid
                        if converged(sol)
                            c_candidate = sol.zero[1]
                            l_candidate = sol.zero[2]
                            w_candidate = sol.zero[3]
                            r_candidate = sol.zero[4]
                            if c_candidate <= 0 || l_candidate <= 0 || w_candidate <= 0 || r_candidate <= 0
                                continue  # Skip if any variable is non-positive
                            end
                        else
                            continue  # Skip if solver did not converge
                        end

                        # Compute government consumption
                        g_local = τ * w_candidate * l_candidate

                        # Compute utility
                        utility = log(c_candidate) + ψ * log(g_local) - 0.5 * l_candidate^2

                        # Compute expected value function using interpolation
                        EV = 0.0
                        for τp_idx = 1:length(tau_states)
                            τ_next = tau_states[τp_idx]
                            π_ττ = Pi_tau[τ_idx, τp_idx]
                            for zp_idx = 1:length(z_states)
                                z_next = z_states[zp_idx]
                                π_zz = Pi_z[z_idx, zp_idx]
                                π_total = π_ττ * π_zz

                                # Interpolate next period's value function
                                kp_idx = searchsortedfirst(k_grid, k_prime)
                                ip_idx = searchsortedfirst(i_grid, i)
                                kp_idx = min(max(kp_idx, 1), k_size)
                                ip_idx = min(max(ip_idx, 1), i_size)

                                V_interp = interpolate((k_grid, i_grid), V[:, :, τp_idx, zp_idx], Gridded(Linear()))
                                V_next = V_interp(k_prime, i)

                                EV += π_total * V_next
                            end
                        end

                        # Total value
                        total_value = utility + β * EV

                        # Update value function if higher utility is found
                        if total_value > V_max
                            V_max = total_value
                            c_opt = c_candidate
                            l_opt = l_candidate
                            w_opt = w_candidate
                            r_opt = r_candidate
                            i_opt = i
                            kprime_opt = k_prime
                        end
                    end

                    # Update the value function and policy functions
                    V_new[k_i, i_prev_i, τ_idx, z_idx] = V_max
                    c_policy[k_i, i_prev_i, τ_idx, z_idx] = c_opt
                    l_policy[k_i, i_prev_i, τ_idx, z_idx] = l_opt
                    w_policy[k_i, i_prev_i, τ_idx, z_idx] = w_opt
                    r_policy[k_i, i_prev_i, τ_idx, z_idx] = r_opt
                    i_policy[k_i, i_prev_i, τ_idx, z_idx] = i_opt
                    kprime_policy[k_i, i_prev_i, τ_idx, z_idx] = kprime_opt

                    # Update sup norm
                    sup_norm = max(sup_norm, abs(V_new[k_i, i_prev_i, τ_idx, z_idx] - V[k_i, i_prev_i, τ_idx, z_idx]))
                end
            end
        end
    end

    # Check for convergence
    println("Iteration $iter, Sup Norm = $sup_norm")
    if sup_norm < tol
        println("Converged after $iter iterations.")
        break
    end

    # Update value function
    V = copy(V_new)
end

# Compute the policy function
println("Policy function computed.")



Starting Value Function Iteration...


LoadError: InterruptException:

Due to very high number of states, the algorithm did not converge and had to be interrupted. This is the case even under the simplifying assumption that the aggregate states are constant.